include/deal.II/lac/petsc_solver.h

```diff
@@ -141,21 +148,53 @@ namespace PETScWrappers
     void
     initialize(const PreconditionBase &preconditioner);
 
+    /**
+     * Return the PETSc KSP object.
+     */
+    KSP
+    petsc_ksp();
+
+    /**
+     * Conversion operator to gain access to the underlying PETSc type. If you
+     * do this, you cut this class off some information it may need, so this
+     * conversion operator should only be used if you know what you do.
+     */
+    operator KSP() const;
+
   protected:
+    /**
+     * The PETSc KSP object.
+     */
+    KSP ksp;
+
     /**
      * Reference to the object that controls convergence of the iterative
      * solver. In fact, for these PETSc wrappers, PETSc does so itself, but we
      * copy the data from this object before starting the solution process,
      * and copy the data back into it afterwards.
      */
-    SolverControl &solver_control;
+    SmartPointer<SolverControl, SolverBase> solver_control;
+
+    /**
+     * Utility to create the KSP object and attach convergence test.
+     */
+    void
+    create_with_comm(const MPI_Comm &comm);
 
     /**
      * %Function that takes a Krylov Subspace Solver context object, and sets
      * the type of solver that is requested by the derived class.
      */
     virtual void
-    set_solver_type(KSP &ksp) const = 0;
+    set_solver_type(KSP &) const
+    {}
```

Comment:
We normally move implementations outside the class declaration. Can you do that?

Reply:
Yes.

Reply:
done